crates/hotshot/types/src/traits/election.rs

```diff
@@ -164,17 +164,44 @@ pub trait Membership<TYPES: NodeType>: Debug + Send + Sync {
 
     #[allow(clippy::type_complexity)]
     /// Handles notifications that a new epoch root has been created
-    /// Is called under a read lock to the Membership. Return a callback
-    /// with Some to have that callback invoked under a write lock.
+    /// This function results in a chain of callbacks being returned. It is invoked
+    /// synchronously under a read lock. Return an Option-wrapped callback to run an
+    /// asynchronous handler to execute. This handler returns another Option-wrapped
+    /// callback, which will be executed synchronously under a write lock to save
+    /// results back to the membership.
     ///
-    /// #3967 REVIEW NOTE: this is only called if epoch is Some. Is there any reason to do otherwise?
+    /// Example usage:
+    /// ```
+    ///    // Do some synchronous work first inside of the read lock
+    ///    Some(Box::new(move || {
+    ///        Box::pin(async move {
+    ///            // We're now in the async context between the two locks.
+    ///            // Do some async work here, like storing the stake table
+    ///
+    ///            // Return a callback to be executed under the write lock
+    ///            Some(Box::new(move |committee: &mut Self| {
+    ///                // Do some synchronous work here, updating the membership
+    ///            }) as Box<dyn FnOnce(&mut Self) + Send>)
+    ///        })
+    ///    }))
+    /// ```
     fn add_epoch_root(
         &self,
```

Comment:
Any chance we can just take a `RwLock<Self>` like in `get_epoch_root` and eliminate the need to have this function return this really complex return type?